[Core] Adding  FindIntersectedGeometricalObjectsWithOBBProcess and other improvements

loumalouomega opened the pull request:
This PR includes several PR that must be approved and merged first and it is only the first step to improve the contact search with OBB.

This PR includes:
- Generalizing FindIntersectedGeometricalObjectsProcess to be sued with Elements and Conditions (not just Elements)
- Creating of the class OBB (Oriented Bounding Box) for better and more fit search than AABB (regular bounding box).  #4264. The tree is still pure AABB, this is used only to compute the intersection. See following image: 

![](https://www.researchgate.net/profile/Gang_Mei2/publication/272093426/figure/fig9/AS:614344521965582@1523482556647/Bounding-volumes-sphere-axis-aligned-bounding-box-AABB-oriented-bounding-box.png)

- FindIntersectedGeometricalObjectsWithOBBProcess process like FindIntersectedGeometricalObjectsProcess that uses OBB to compute the intersections (implicit contact require potential surfaces not necessarilty intersected or in direct contact) #4707
- Enhances the OctreeBinary #4262
- Correcting files affected indirectly by the changes
- Tests

This PR includes the following PR:

- #4264
- #4262
- #4707

loumalouomega committed:
[Core] Clean up tabulation FindIntersectedElementsProcess cpp test

Replacing tabulation with 4 spaces as style guide requires

loumalouomega committed:
Cleaning up the porcesses too

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Several WIP

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
WIP several clean ups

loumalouomega committed:
WIP

loumalouomega committed:
Compiles

loumalouomega committed:
Extending to conditions

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[Core] Adding Edges to Line2D2 and Line3D2

This adds the Edge method to Line2D2 and Line3D2 that wasn't defined

loumalouomega committed:
Adding first test for 2D condition

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Adding Parameter constructor

loumalouomega committed:
Adding non intersected test

loumalouomega committed:
Toward extended bounding box

loumalouomega committed:
[Core] Adding missing Faces() method to the Triangle3D3 and Quadrilateral3D4

This PR adds the missing methods  to the Triangle3D3 and Quadrilateral3D4

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[Core] Small correction + test for Geometry Bounding Box

THe code can be simplified (first node is already assigned, so can be skipped). Adding non existing test

loumalouomega committed:
OOB WIP

loumalouomega committed:
Comment

loumalouomega committed:
Cleaning

loumalouomega committed:
Separating between source and header

loumalouomega committed:
Fix

loumalouomega committed:
[skip ci] Fix

loumalouomega committed:
[skip ci] Separating FindIntersectedGeometricalObjectsWithOBBProcess

loumalouomega committed:
[skip ci]  Impoving Bounding box

loumalouomega committed:
[skip ci] Intersection with OBB

loumalouomega committed:
[skip ci] Some chnages OBB

loumalouomega committed:
[skip ci] Corrections

loumalouomega committed:
[skip ci] More than one axis

loumalouomega committed:
[skip ci] Debug info

loumalouomega committed:
[skip ci] Advances

loumalouomega committed:
Adding test

loumalouomega committed:
[skip ci] Better test

loumalouomega committed:
[skip ci] Some fixes

loumalouomega committed:
Adding 3D test (WIP)

loumalouomega committed:
[skip ci] Improving check

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[skip ci] 3D rotation

loumalouomega committed:
Correct test and coordinates

loumalouomega committed:
[skip ci] More test

loumalouomega committed:
[Core] Adding HasIntersection method for Quadrilateral3D4N

HasIntersection method was missing in Quadrilateral3D4N. Test added

loumalouomega committed:
Extend check

loumalouomega committed:
[skip ci] Moving inverted rotation matrix for performance

loumalouomega committed:
[skip ci] Adding GetModelPart2

loumalouomega committed:
[skip ci] Adding advance debugging options

loumalouomega committed:
[skip ci] Minor

loumalouomega committed:
Test working

loumalouomega committed:
Computing OBB in 3D

loumalouomega committed:
3D test (WIP)

loumalouomega committed:
Missing abs

loumalouomega committed:
[skip ci] More consistency

loumalouomega committed:
Check

loumalouomega committed:
Minor

loumalouomega committed:
[skip ci] Adding offset and scale

loumalouomega committed:
[skip ci] SetScaleFactor and SetOffset

loumalouomega committed:
[skip ci] Minor

loumalouomega committed:
[skip ci] Becoming the parameters into member variable

loumalouomega committed:
[skip ci] Adding GetScaleFactor and GetOffset

loumalouomega committed:
[skip ci] Update the definition of BB

loumalouomega committed:
[skip ci] Correction

loumalouomega committed:
Commenting include

loumalouomega committed:
Exporting to python

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Change name

loumalouomega committed:
Correct

loumalouomega committed:
Changing name of the class

loumalouomega committed:
Changing name class

loumalouomega committed:
Update header

loumalouomega committed:
Making it virtual

loumalouomega committed:
Making more modular the base class

loumalouomega committed:
Making public

pooyan-dadvand commented:
@loumalouomega Are you sure that this would improve your performance? And if it does, how much?

loumalouomega commented:
> @loumalouomega Are you sure that this would improve your performance? And if it does, how much?

I haven't finished yet the implementation. In any case is more a problem of robustness than performance (for implicit cases).

pooyan-dadvand commented:
Would you please explain more why this is a robustness problem? Consider that the AABBs are the most robust structure we know

loumalouomega commented:
> Would you please explain more why this is a robustness problem? Consider that the AABBs are the most robust structure we know

Copy/pasting myself from #4262:

- Explicit contact: You check that the geometries are actually intersected (due to the small time increment this will not affect the final solution)
- Implicit contact: you must predict potential contact pairs. Direct intersection is not enough, the geometries must be "extruded" in a offset. I could use directly the AABB standard bounding box, but in terms of robustness is not much better than the current approach which is the sphere. The OBB is more thigh so gives a better prediction of the potential pairs.

loumalouomega committed:
Minor

loumalouomega committed:
Adding constructor requested by @pooyan-dadvand

loumalouomega committed:
Adding SeparatingAxisTheoremHasIntersection

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

pooyan-dadvand commented:
Two questions:
1. I see that OBB gives less cell to check for intersection but the check is more expensive. I imagine that OBB should be better but do you have any comparison in time between AABB/AABB and AABB/OBB contact search?
2. Still you have not answered my question about robustness problem and how OBB would improve it

loumalouomega commented:
> Two questions:
> 
>     1. I see that OBB gives less cell to check for intersection but the check is more expensive. I imagine that OBB should be better but do you have any comparison in time between AABB/AABB and AABB/OBB contact search?
> 
>     2. Still you have not answered my question about robustness problem and how OBB would improve it

Both questions are related. OBB gives a smaller cell, then the number of resulting potential pairs will be smaller (we want to have the smallest number of potential pairs possible to avoid over constraint the problem, and therefore ill conditionate the problem).

I can not give search times right now, this is a transition PR, and the final PR focused on contact search will came later (during this week)

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Removing new methods

loumalouomega committed:
Removing dependency of new methods

loumalouomega committed:
More clean up

loumalouomega committed:
Minor typo

loumalouomega committed:
Clean up

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Moving to protected variables

loumalouomega committed:
Remove unused parameters

loumalouomega committed:
Clean unused parameters

loumalouomega committed:
Missing const

loumalouomega committed:
Missing Info, PrintInfo, PrintData

loumalouomega committed:
Avoid comparison compilation error in Travis

loumalouomega committed:
Adding check

loumalouomega committed:
Improving debug info

loumalouomega committed:
Adding intersection type to the process

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
More debug info

loumalouomega committed:
Adding more complex case

loumalouomega committed:
Extending BB

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Some clean up

loumalouomega committed:
Improving HasIntersection from OctreeBinary

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Moving to protected

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Revert authorship

loumalouomega committed:
Adding geometry constructors

loumalouomega committed:
Using constructor to reduce code duplication

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Missing KRATOS_API

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Missing template argument

loumalouomega committed:
Fix

loumalouomega committed:
Moving FastCrossProduct to MathUtils

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Solving conflict

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Deleting assigment operator

loumalouomega committed:
Default destructor

loumalouomega committed:
Moving to this class the parameters

loumalouomega committed:
Removing Parameters

loumalouomega committed:
Reducing code duplication

loumalouomega committed:
More deterministic test

loumalouomega committed:
Updating process for OBB

loumalouomega committed:
Update name of model parts

loumalouomega committed:
More code reduction

loumalouomega committed:
Updating name for OBB

loumalouomega committed:
Revert "Updating name for OBB"

This reverts commit 87bec052f8c94a8d1976c75085d44a4204c654ef.

loumalouomega committed:
Updating name for base class

loumalouomega committed:
Revert "Updating name for base class"

This reverts commit 43e35a1511406c2af13601d7e93b7d7baac63d4b.

loumalouomega committed:
Correct name

loumalouomega committed:
Fix

loumalouomega committed:
Windows compilation

loumalouomega committed:
Linux compilation issue

loumalouomega committed:
Adding template argument to octree_binary

loumalouomega committed:
Generalizing combinations

@rubenzorrilla requested

loumalouomega committed:
Adding test

loumalouomega committed:
Extending to OBB process

loumalouomega committed:
Expose to python (I)

loumalouomega committed:
Expose to python (II)

loumalouomega committed:
Reverting python exposition

loumalouomega committed:
Fix

loumalouomega committed:
Reverting math_utils

loumalouomega committed:
Using CrossProduct

rubenzorrilla committed:
Delete

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

rubenzorrilla committed:
Delete

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Adding entities utilities

loumalouomega committed:
Reducing  explicit templates

loumalouomega committed:
Reverting specifications

loumalouomega committed:
Reverting changes

loumalouomega committed:
Update test

loumalouomega committed:
Refactoring without templates (I)

loumalouomega committed:
Refactoring without templates (II)

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Missing override

loumalouomega committed:
Adding direct intersection (I)

loumalouomega committed:
Adding direct intersection (II)

loumalouomega committed:
Reverting changes geometries

rubenzorrilla committed:
Missing const

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

loumalouomega committed:
@rubenzorrilla comment

loumalouomega committed:
Merge master

loumalouomega committed:
[Core] Moving flag from Condition/Element to GeometricalObject

Basically that. The flag is common between the geometries and conditions, but currently both derive from Flag instaed of bein included on the GeometricalObject

loumalouomega committed:
Reducing overhead of Element/Condition

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Serializer correction

loumalouomega committed:
Exposing Flags to python

loumalouomega committed:
@jcotela commentaries + doxygen

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Adding local flags

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Adding flags

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega commented:
This PR now just adds new features, not modifying previous code

loumalouomega commented:
Ping

pooyan-dadvand commented on kratos/processes/find_intersected_geometrical_objects_with_obb_process.cpp:
```diff
@@ -0,0 +1,455 @@
+    ModelPart& rModelPartIntersected,
+    ModelPart& rModelPartIntersecting,
+    const double BoundingBoxFactor,
+    const bool DebugOBB,
```
can be merged with option flag

loumalouomega replied:
OK

pooyan-dadvand commented on kratos/processes/find_intersected_geometrical_objects_with_obb_process.cpp:
```diff
@@ -0,0 +1,455 @@
+    ModelPart& rModelPartIntersecting,
+    const double BoundingBoxFactor,
+    const bool DebugOBB,
+    const OBBHasIntersectionType IntersectionType,
```
This can also be passed by option flags

loumalouomega replied:
OK

pooyan-dadvand commented on kratos/processes/find_intersected_geometrical_objects_with_obb_process.cpp:
```diff
@@ -0,0 +1,455 @@
+            if (std::abs(high[i] - low[i]) < mBoundingBoxFactor) {
+//                 low[i] -= mBoundingBoxFactor;
+                high[i] += mBoundingBoxFactor;
+            }
```
I would suggest to inflate the bounding box by min and max and not only max point. This gives robustness issues specially when min point is origin

loumalouomega replied:
OK

loumalouomega replied:
For some reason this doesn't appear as updated/outdated

pooyan-dadvand reviewed (commented):
Too many debugging ifs. Is it possible to have a print mesh method which writes the OBBs only in debug?

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Reordering flags (start on 0)

loumalouomega committed:
Using flags

loumalouomega committed:
Updating test

loumalouomega committed:
Removing comment

loumalouomega committed:
Merge branch 'master' into core/find_intersected_geometrical_objects_with_obb_process

loumalouomega committed:
Lower extend

loumalouomega commented:
This is ready @pooyan-dadvand

pooyan-dadvand reviewed (approved):
:+1:

loumalouomega commented:
Thanks
